overlord/exportstate/exportstate.go

```diff
@@ -0,0 +1,450 @@
+// exist no error is reported. This is because this function is most often
+// called from link-snap where it runs unconditionally but most snaps do not
+// have any content to export and the symlink would be dangling.
+func setCurrentSubKey(primaryKey, subKey string) error {
+	pathName := currentSubKeySymlinkPath(primaryKey)
+	if err := osutil.AtomicSymlink(subKey, pathName); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("cannot set current subkey of %q to %q: %v", primaryKey, subKey, err)
+	}
+	return nil
+}
+
+// removeCurrentSubKey removes the "current" symlink for the given primary key.
+func removeCurrentSubKey(primaryKey string) error {
+	if err := os.Remove(currentSubKeySymlinkPath(primaryKey)); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("cannot remove current subkey of %q: %v", primaryKey, err)
+	}
+	return nil
+}
+
+// ManifestKeys returns the (primaryKey, subKey) tuple to use as the current
+// provider of all the export sets of a given snap. The returned subKey may be
+// empty, indicating that given snap has no current revision.
+func ManifestKeys(st *state.State, instanceName string) (primaryKey string, subKey string, err error) {
+	switch instanceName {
+	case "core", "snapd":
+		primaryKey, subKey, err = manifestKeysForSnapdOrCore(st)
+		if err != nil {
+			return "", "", err
+		}
+	default:
+		info, err := snapstateCurrentInfo(st, instanceName)
+		if _, ok := err.(*snap.NotInstalledError); err != nil && !ok {
+			return "", "", err
+		}
+		if info == nil || info.Broken != "" {
+			primaryKey, _ = snap.SplitInstanceName(instanceName)
+			return primaryKey, "", nil
+		}
+		primaryKey, subKey = manifestKeysForRegularSnap(info)
+	}
+	return primaryKey, subKey, nil
+}
+
+// manifestKeysForRegularSnap returns the subkey to use as the current provider of
+// all the export sets of a given snap. The given snap info must be the active
+// revision of the given snap.
+//
+// This function does not work for core and snapd.
+func manifestKeysForRegularSnap(info *snap.Info) (primaryKey string, subKey string) {
+	if info.SnapName() == "core" || info.SnapName() == "snapd" {
+		panic("internal error, cannot use manifestKeysForRegularSnap with core or snapd")
+	}
+	primaryKey = info.SnapName() // Instance key goes to subKey
+	if info.InstanceKey == "" {
+		subKey = info.Revision.String()
+	} else {
+		subKey = fmt.Sprintf("%s_%s", info.Revision.String(), info.InstanceKey)
+	}
+	return primaryKey, subKey
+}
+
+// manifestKeysForSnapd returns the sub key for all export sets from the snapd snap.
+func manifestKeysForSnapd(info *snap.Info) (primaryKey string, subKey string) {
+	return "snapd", info.Revision.String()
+}
+
+// manifestKeysForCore returns the sub key for all export sets from the core snap.
+//
+// The return value is the custom core_$revision string.
+func manifestKeysForCore(info *snap.Info) (primaryKey string, subKey string) {
+	return "snapd", fmt.Sprintf("core_%s", info.Revision)
```

Comment:
Is the primaryKey deliberately set to "snapd" for "core" (I guess it is and I think it makes sense, just double-checking)?

Reply:
Yes